GridView

Koerty opened the pull request:
Integration of the lastest version of Jacofk20's GridView.

![screenshot](https://user-images.githubusercontent.com/24305945/28986464-63b46b1c-7967-11e7-963b-04419d09ca21.png)
(screenshot of the GridView shipped with this pull requests, using material theme from @lilbud )

The only difference is I've enabled system switch with LeftShoulder and RightShoulder for all the game views (even if it's only displayed on gridview's system help atm). The reason is for now if you enable the gridview you have it for every system, but I plan on adding more control for the user on views (maybe allow him to choose a different view for every system individually, or a second automatic option but choosing automatically between grid and basic view instead of detailed and basic view like it is atm, idk yet I need to think more about it). So, only allowing him to switch with LeftShoulder and RightShoulder on gridview kinda break the flow every time he encounter another kind of view (idk why Jacobfk20 did this it's kinda weird).

I will start working on better integration, refactoring, optimization, more features for gridview ..... but will ship them separatly in different pull requests, since this one is already big enough.

LodanZark commented:
Looks great, definitely would be better remove the game name under box part and put it in the right side section between console logo and rating display to avoid issues with games with long names and make it look more "flat" and clean

Koerty commented:
@LodanZark This comment is irrelevant. The material theme choose to hide the title of the game, displayed at the top of the screen by default, by setting his position to [1,1]

dudleydes commented:
With Jacobfk20's Release-Beta branch, it is possible to set view modes for individual systems. In game view, if I press the select button on my controller, then `System UI Setting`, then I get options to select view mode (see the last image in this [album](https://imgur.com/a/NQYjV)) and grid size. These settings are saved in `'es_systems.cfg` with `<viewmode>` and `<gridsize>` tags as well as in `es_systems-extra.cfg`.

Koerty commented:
@dudleydes This isn't part of this pull request but will be added in the future. I need to search/think/discuss more about it (check the big paragraph right under the screenshot in my first comment)

I may be wrong, but for me allowing the user to change the view for every system is a bit overkill.

A more conveniant way would be to allow the user to choose one view for system which have metadatas and another one for system which don't have metadatas, a bit like the "automatic" view do at the moment.

Instead of setting view type to : Basic or Detailed or Automatic or Grid
you would have 2 options in the menu
View type for systems without metadatas : Basic or SimpleGrid
View type for systems with metadatas : Basic or Detailed or SimpleGrid or DetailedGrid

But as I said, this need more research/thinking/discussion about user usage and more development, so I will ship it in another pull request in the next days/weeks.

Koerty commented:
@joolswills @pjft @jrassa Does any of you have some free time soon to take a first look at this PR ? I really want to do more but I'm quickly loosing motivation because I feel like this PR is going to sit here forever

pjft commented:
Oh. It wasn't clear to me at least that it was ready for review. The last
several updates were you pushing new commits, so I assumed you were still
developing.

For the sake of best use of time, it's better to review a PR when it's
actually stable and ideally final unless otherwise requested. Otherwise we
may be reviewing things that will be changed. Anything in particular you'd
want feedback on at this stage? Or is it ready from your point of view?

I can't promise much in the next few weeks, but if I do have the chance
I'll read through the code and give some feedback in the coming days. I'll
say, though, that views and themes are not my area of expertise, so
regardless of my review I'd be happier if @jrassa or @zigurana had the
chance as well, as I may miss some obvious things there.
On Sun, 13 Aug 2017 at 13:30 Koerty <notifications@github.com> wrote:

> @joolswills <https://github.com/joolswills> @pjft
> <https://github.com/pjft> @jrassa <https://github.com/jrassa> Does any of
> you have some free time soon to take a first look at this PR ? I really
> want to do more but I'm quickly loosing motivation because I feel like this
> PR is going to sit here forever
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/203#issuecomment-322039311>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7VALAn_zQaPWhrC3RFdfOZ7jE0u3ks5sXux4gaJpZM4OuIr4>
> .
>

Koerty commented:
@pjft I think it's ready for a first review, this is the grid view as Jacofk20 designed it with minor modifications for better integration in this repository. I'm working on upgrades for the grid view but will ship them in different pull requests so it will be easier for you to see what got modified in the process.

jrassa commented on es-core/src/ThemeData.h:
```diff
@@ -73,6 +73,19 @@ ThemeException& operator<<(ThemeException& e, T appendMsg)
 	return e;
 }
 
+class ThemeExtras : public GuiComponent
```
The `ThemeExtras` class was removed in the time since this grid view implementation was originally done.  This change needs to be reverted.

jrassa commented on es-app/src/views/gamelist/ISimpleGameListView.cpp:
```diff
@@ -8,7 +8,7 @@
 
 ISimpleGameListView::ISimpleGameListView(Window* window, FileData* root) : IGameListView(window, root),
-	mHeaderText(window), mHeaderImage(window), mBackground(window)
+	mHeaderText(window), mHeaderImage(window), mBackground(window), mThemeExtras(window)
```
The `ThemeExtras` class was removed in the time since this grid view implementation was originally done.  This change needs to be reverted.

jrassa commented on es-app/src/views/gamelist/ISimpleGameListView.cpp:
```diff
@@ -35,20 +36,7 @@ void ISimpleGameListView::onThemeChanged(const std::shared_ptr<ThemeData>& theme
 	mHeaderImage.applyTheme(theme, getName(), "logo", ALL);
 	mHeaderText.applyTheme(theme, getName(), "logoText", ALL);
 
-	// Remove old theme extras
```
The `ThemeExtras` class was removed in the time since this grid view implementation was originally done.  This change needs to be reverted.

jrassa replied:
The `ThemeExtras` class was removed in the time since this grid view implementation was originally done.  This change needs to be reverted.

jrassa commented on es-app/src/views/gamelist/ISimpleGameListView.h:
```diff
@@ -32,7 +32,7 @@ class ISimpleGameListView : public IGameListView
 	ImageComponent mHeaderImage;
 	ImageComponent mBackground;
 
-	std::vector<GuiComponent*> mThemeExtras;
```
The `ThemeExtras` class was removed in the time since this grid view implementation was originally done.  This change needs to be reverted.

Koerty commented:
@jrassa Sorry for the noob question but how will I notify you that the requested changes were made ? Is there any particular tag I should add to my commit or anything like that ?

Koerty commented:
@jrassa Sorry for that changes on ThemeExtras, I reverted it back and added (back ?) the Zindex, did you added it or Jacobfk20 removed it ?

I also removed "big" and "small" grids view in previous commit. In Jacobfk20's version the user could choose for each system if he want a bigger or smaller grid. With the code from this PR there is not this entry in the menu so this code is useless for now. I will consider adding this entry in the menu, or another way to change grid size for the user in the futur, or just not implement it at all, as we agreed on my other pull request.

Thanks for this first review by the way, I hope you are satisfied with this changes ;)

jrassa commented on es-core/src/components/IList.h:
```diff
@@ -36,15 +37,14 @@ struct ScrollTierList
 // default scroll tiers
 const ScrollTier QUICK_SCROLL_TIERS[] = {
 	{500, 500},
-	{2000, 114},
```
This should be reverted as well.

Should be:
```cpp
// default scroll tiers
const ScrollTier QUICK_SCROLL_TIERS[] = {
	{500, 500},
	{2000, 114},
	{4000, 32},
	{0, 16}
};
const ScrollTierList LIST_SCROLL_STYLE_QUICK = { 4, QUICK_SCROLL_TIERS };

const ScrollTier SLOW_SCROLL_TIERS[] = {
	{500, 500},
	{0, 200}
};
const ScrollTierList LIST_SCROLL_STYLE_SLOW = { 2, SLOW_SCROLL_TIERS };
```

jrassa commented:
@Koerty Just as an FYI, eventually we will want you to squash your code down to a single commit.  That is the way that RetroPie prefers it to, so that it keeps the commit log cleaner.

Koerty commented:
@jrassa Ok I reverted the changes on ScrollTierList and squashed to a single commit. I hope the squash is ok I never did that before, I have modifications from earlier PRs in my commit, like modifications on PowerSaver.cpp, I think it's because my GridView branch is based on my local master branch and I merged upstream/master to get your last change about rotation and origin support.

I hope that's not a big deal, if that's so, I can open a new PR since I did a cherrypick of all my commits on GridView and saved them on another branch called GridViewSave before squashing.

zigurana commented on data/ResourceUtil.cpp:
```diff
@@ -1,104 +1,106 @@
 #include "Resources.h"
 
-const size_t res2hNrOfFiles = 46;
+const uint32_t res2hNrOfFiles = 47;
```
Regarding this change of resources, there are two questions I have:
1. Why do we need a hardcoded default image for images without an icon? To me, that seems something that a theme can (and should) define by itself.
2. Why is the type changed from `size_t` to `uint32_t`? Is that the result of a newer/different version of res2h?

As I see it now, this re-run of res2h now causes a lot of changed files (48 files out of 86 in total), but not actually a change in program behavior. Similar to style changes without a need, why are you doing this?

zigurana replied:
NB: [I see](https://github.com/HorstBaerbel/res2h/commit/c7088db9b4d706a318e2aaa6a84f092caa43f1af) that rs2h was updated about 7 months ago, to enable 32bit and 64bit archives, which causes this type change.
I guess we need to make this transition at some point anyway, but it sure makes this review a bit unwieldy. My preference would be to do this in another, separate commit altogether, as it in itself has nothing to do with GridView.

Koerty replied:
1. It's completly normal to have a hardcoded default image for games without boxart

The theme can also add a custom image for games without boxart with the following syntax as specified in the GRIDTHEME.md : 
```xml
// Replace the missing boxart texture
<image name="missing_boxart">
	<path>new_path</path>
</image>
```

2. You answered it by yourself.

The thing is the default image for games without boxarts is part of the gridview, should I also include it in the other pull request with the rs2h update ?

Koerty replied:
By the way, as I stated before, the goal of this first pull request is to give an "as it" version of the gridview, I can add small changes like the xml modifier to make it fancy, but if I had to add bigger changes like allow themes to change the image for games without boxarts (if it wasn't already here), I would prefer to do it in separates pull requests.

zigurana commented on GRIDTHEME.md:
````diff
@@ -0,0 +1,102 @@
+```
+
+Setup for the grid as a whole.
+```
````
You can add the xml modifier to the code-block, if you want to make this fancy: ` ```xml `

zigurana commented on GRIDTHEME.md:
```diff
@@ -0,0 +1,102 @@
+GridView Theme
```
I am assuming some of this will change as a result of our discussion on the theming syntax?

Koerty replied:
Not today but in a near future there is a possibility that this will change, I still need to think more about it.

zigurana commented on THEMES.md:
```diff
@@ -517,6 +517,10 @@ Can be created as an extra.
 	- The image will be resized as large as possible so that it fits within this size and maintains its aspect ratio.  Use this instead of `size` when you don't know what kind of image you're using so it doesn't get grossly oversized on one axis (e.g. with a game's image metadata).
 * `origin` - type: NORMALIZED_PAIR.
 	- Where on the image `pos` refers to.  For example, an origin of `0.5 0.5` and a `pos` of `0.5 0.5` would place the image exactly in the middle of the screen.  If the "POSITION" and "SIZE" attributes are themable, "ORIGIN" is implied.
+* `rotation` - type: FLOAT.
```
These changes are the result of an[ earlier commit](https://github.com/RetroPie/EmulationStation/commit/2377175a1995ae4cf707bce90005c14028db75ed), did you rebase your branch before squashing your commits, or did you merge in these changes manually? This way, we are reviewing those changes as well, while they are already part of the code-base.

Koerty replied:
Should I rebase on upstream/master or on my local master ?

Koerty replied:
Otherwise I saved my changes on another branch named GridViewSave, what if I delete this branch and create a new branch with the same name (GridView) ? Will this pull request be linked to my new GridView branch ?

jrassa replied:
@Koerty don't delete this your branch or close this PR.  I will get you some instruction on how you can update your GridView branch.

zigurana replied:
Ok, so I'm far from a Github Guru, but I understand it as follows:
Any change you now push to this branch, will automatically update this PR.
Now, if your remote for the retropie branch is called RP, you can do:
```
git fetch RP
git rebase RP/master GridView
```
Which should start a rebase activity, which will probably result in conflicts for those files you have currently merged explictly (screensaver/powersaver etc). Resolving those conflicts should be trivial, as these changes are identical. They are both in the same files, at the exact same locations though, so git will see them as overlapping edits, I think, (if not then the whole thing is painless anyways).
After you've resolved the conflicts, and added them to the commit, the rebase will finalize, resulting in a single commit with just your changes.

Good luck, let me know how that goes!

zigurana commented on es-app/src/EmulationStation.h:
```diff
@@ -4,10 +4,10 @@
 #define PROGRAM_VERSION_MINOR        5
-#define PROGRAM_VERSION_MAINTENANCE  1
-#define PROGRAM_VERSION_STRING "2.5.1rp"
+#define PROGRAM_VERSION_MAINTENANCE  2
```
This is for the Repo maintainer to bump.

zigurana commented on es-app/src/SystemScreenSaver.cpp:
```diff
@@ -4,6 +4,7 @@
 #endif
 #include "components/VideoVlcComponent.h"
 #include "platform.h"
+#include "PowerSaver.h"
```
Again, these changes are not part of GridView feature improvements, but likely the result of manual merging of commits to RetroPie/master after your branching point.
Please rebase the branch, to clean these up.

(Git is a pain, I know).

Koerty commented:
@zigurana Jesus mother f***ing christ, it worked ! You can now review my PR in peace with only what's needed. You can find the ressources update thanks to Res2h on my other PR

Koerty commented:
I had to do a bunch of cherrypick/revert commit/rebase .. but it worked ! Sorry for the inconvenience

zigurana commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -90,27 +140,246 @@ class ImageGridComponent : public IList<ImageGridData, T>
+	Eigen::Vector2i mDesiredGridSize;
+
+	const int MAX_TEXTURES = 80;			// The maximum amount of images that can be loaded at once
+	const int CURSOR_RANGE = 34;			// How many images will be loaded around the cursor [ Cursor will be center of range ]
```
wait, now we have another default value, 34 rather than 12?

zigurana commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -90,27 +140,246 @@ class ImageGridComponent : public IList<ImageGridData, T>
+	Eigen::Vector2f mMargin;
+	Eigen::Vector2i mDesiredGridSize;
+
+	const int MAX_TEXTURES = 80;			// The maximum amount of images that can be loaded at once
```
Where does the magic number come from?

Koerty replied:
From my magical hat ... I have no idea for this one

Koerty replied:
This look like a security check to prevent leaks when the dynamic image loader load too many images in very specific case.

zigurana commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -90,27 +140,246 @@ class ImageGridComponent : public IList<ImageGridData, T>
+}
+
+template<typename T>
+void ImageGridComponent<T>::unloadTextures(bool unloadAll) {
```
maybe make this method save / cheap to call when empty?

Koerty replied:
What do you mean ?

zigurana replied:
hmm, I don't remember :-/
I think my reasoning was something like, we call this from several locations, but we do not want to include calling-criteria at every calling-instance, so maybe just always call, even if the list is empty to begin with. But there is nothing to really optimize here anyways, so...
Never mind!

zigurana commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -90,27 +140,246 @@ class ImageGridComponent : public IList<ImageGridData, T>
+	if (mLoadedTextureList.size() > MAX_TEXTURES) unloadTextures();
+
+	// if cursor is getting close to the ends of range; update
+	if (getCursorIndex() > mCursorRange.max - 5 && getCursorIndex() < mCursorRange.min + 5) updateLoadRange();
```
why 5?

zigurana commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -158,18 +439,76 @@ void ImageGridComponent<T>::render(const Eigen::Affine3f& parentTrans)
+
+	listRenderTitleOverlay(trans);
+
 	GuiComponent::renderChildren(trans);
```
Do all children need to be rendered on top of the tile? Also, do we need to support z-index for any of this?

Koerty replied:
I am not sure about this

zigurana commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -158,18 +439,76 @@ void ImageGridComponent<T>::render(const Eigen::Affine3f& parentTrans)
 }
 
 template<typename T>
+void ImageGridComponent<T>::applyThemeToChildren(const std::shared_ptr<ThemeData>& theme) {
```
yeah, so to me this look like a regular applyTheme, not the applyTheme to the children of the ImageGridComponent...

Koerty replied:
Changed to regular applyTheme

zigurana commented:
My next step for this weekend is to actually try and build this on my windows PC and on the PI, so I can test it. See if I can make it break. Do you maybe have a test-theme I could use?

Koerty commented:
> Wow! :+1:
This was a tremendous job of you, good work!
This is a big feature, and the fact alone that I was able to get through these 3000 lines you added within 3 hours, speaks volumes of your style and readibility!

Wow wow wow let's stop right there ! I don't like taking credits for other people's work, the grid view as you see it was developed by @Jacobfk20 (this is literally the first line of this pull request, right above the screenshot)

I just did the integration, changed the minimal amount of code to make it work and deliver an "as it" version. The goal was to add more stuff in next PR (the applyTheme and onThemeChanged functions as you noticed, clean unused code, apply the scroll sound ...)

What should we do now ? I don't want to change too much code in this PR but I can't just ignore all the questions and suggestions you made.

lilbud commented:
Zigurana: I'm not exactly sure what is happening with the state of all of this. So here is what I say, if the old gridview syntax is still in place, use my Material or Modern TV themes. If new syntax has been added, you are SOL my friend.

jrassa commented:
> I just did the integration, changed the minimal amount of code to make it work and deliver an "as it" version. The goal was to add more stuff in next PR (the applyTheme and onThemeChanged functions as you noticed, clean unused code, apply the scroll sound ...)
> 
> What should we do now ? I don't want to change too much code in this PR but I can't just ignore all the questions and suggestions you made.

@Koerty you have expressed that you would like this PR to be accepted relatively "as is" before making additional changes.  However, you have to understand that the feedback you are getting is the same feedback we would be giving jacobfk20 if he had submitted this himself, and we would have expected him to address it before it would be merged.  We are not going to just accepted this as is with the understanding that feedback will be addressed in a future PR.

Koerty commented:
@jrassa The thing is I can't do a clean and full intergration without completly changing grid theming syntax. That's why I wanted to ship that in a separate pull requests

Koerty commented:
Plus this PR is already really heavy, both in term of code and comments, nobody but me have read the 3000 lignes of codes and 70 comments ...

zigurana commented:
... and me, yesterday-evening.

I understand it's a big job, which is why I am so grateful you are picking it up.
One of the reasons I believe Jacob never submitted a PR himself was that it was not finished yet. Part of that is making it work functionally, but another is getting the code in shape. 
We all know that getting to really understand someone else's code is almost as much work as writing your own from scratch. That is why, even if you might not have started this effort, finishing it also makes you a significant owner of this. This is also your work now, be proud of that!

Please understand that you are really almost there, most of my comments are minor ones, easy to fix. Also, it does not need to contain all the extra features you might have envisioned for it. It just needs to be of sufficient quality on its own, and not requiring fixes from the start.
Adding additional functionality such as the sounds, or a different way to treat the xml, or whatever you want, **those** can come in a later PR.

Koerty commented:
@zigurana I'm not sure you checked it all, since you ask questions like "what theme can I use to test it ?" and the answer is in the first comment of this PR.

@jrassa @zigurana I think you still don't get it, my question is should I completly change the grid theme syntax here ? Because some suggestions here are relative to how themes are applied (functions onThemeChanged, applyTheme, applyThemeToChildren ...) and I can't change it so it completly respect the current way of doing it without also completly changing the grid theming syntax.

zigurana commented:
Well, I did say 'read', not 'learn by heart' in the three hours I invested in reviewing your PR. But no worries, I will find a way to test it.

Re. XML syntax: why don't you bring a proposal to the table (either here, or in the syntax discussion) that incorporates the issues of the `onThemeChanged()` and `applyTheme()` methods? If you feel that impacts the XML parsing directly, that would be a strong argument to go for a particular solution. 
Now, both the code and the syntax are somewhat unclear, but if the former has consequences for the latter, that makes it easier to decide what we need to do with the syntax.

Koerty commented:
Because I need to think more about it, sadly I won't have much free times in the incoming weeks so I wanted to finish this PR first.

Koerty commented:
About testing there is no easy way to test it atm, either checkout my gridview branch and remove the two "ROTATION" attribute from GridGameListView.cpp, or checkout a copy of this repository's master branch and merge my GridView branch into it.

Koerty commented on es-core/src/components/IList.h:
```diff
@@ -244,6 +292,12 @@ class IList : public GuiComponent
 			scroll(mScrollVelocity);
 	}
 
+	void listRenderFileTitle(const Eigen::Affine3f& trans) {
```
Nowhere. Removed

Koerty commented on es-core/src/AnimationUtil.cpp:
```diff
@@ -0,0 +1,83 @@
+	int tab = (int)(getColorDifference(eb, b) / endFrame);
+	int tae = (int)(getColorDifference(ea, a) / endFrame);
+
+	// Get per frame itteration
```
Fixed

Koerty commented on es-core/src/AnimationUtil.cpp:
```diff
@@ -0,0 +1,83 @@
+#include "AnimationUtil.h"
+#include "resources/ResourceManager.h"
+#include "platform.h"
+#include "Log.h"
```
Removed all of them but AnimatioUtil.h

Koerty commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -76,12 +121,17 @@ class ImageGridComponent : public IList<ImageGridData, T>
 
 	Eigen::Vector2i getGridSize() const
 	{
+		if (mDesiredGridSize.x() > 0) return mDesiredGridSize;
```
Done

Koerty commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -76,12 +121,17 @@ class ImageGridComponent : public IList<ImageGridData, T>
 
 	Eigen::Vector2i getGridSize() const
```
Done

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -49,11 +154,281 @@ void GridGameListView::launch(FileData* game)
+	bDefaultTheme = !theme->getElement("grid", "md_grid", "text");
+
+	// If the selected theme contains no grid view elements, create a default theme.
+	if (bDefaultTheme) {
```
Default settings are now sets in the a function getDefaultElements are always applied (before theme grid settings if he have any)

Koerty replied:
EDIT : actually I reverted that change and made it another way : the default settings are all set in the constructor of the GridGameListView class

Koerty commented:
@jrassa @pjft @zigurana By default the metadatas were hidden in the gridview when the theme don't ask specificly to show them. Should we keep this behaviour or do like we do for the detailed list view: show them all by default and the theme must set the position of the metadata it want to hide to 1:1 ?

Koerty commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -18,26 +38,51 @@ class ImageGridComponent : public IList<ImageGridData, T>
 	bool input(InputConfig* config, Input input) override;
 	void update(int deltaTime) override;
 	void render(const Eigen::Affine3f& parentTrans) override;
+	void applyThemeToChildren(const std::shared_ptr<ThemeData>& theme);
```
I changed that to handle it the same way than TextListComponent

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -49,11 +154,281 @@ void GridGameListView::launch(FileData* game)
+}
+
+void GridGameListView::onFocusLost() {
+	if (mGrid.getEntryCount() > 0)
```
Removed this unecessary check

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -49,11 +154,281 @@ void GridGameListView::launch(FileData* game)
+		return;
+	}
+
+	// ---  DETAILED METADATA THEME ---
```
Yes I double checked that a couple of times. This is the exact same metadata stuff than in the current DetailedGamelistView

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -49,11 +154,281 @@ void GridGameListView::launch(FileData* game)
+
+	// If the selected theme contains no grid view elements, create a default theme.
+	if (bDefaultTheme) {
+		mBackground.applyTheme(theme, "basic", "background", ALL);
```
I removed that reference

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -49,11 +154,281 @@ void GridGameListView::launch(FileData* game)
+
+	using namespace ThemeFlags;
+	mTitle.applyTheme(theme, getName(), "md_title", ALL);
+	mGrid.applyTheme(theme, getName(), "md_grid", ALL);
```
Done

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -20,27 +88,64 @@ FileData* GridGameListView::getCursor()
-	{
-		mGrid.add((*it)->getName(), (*it)->getThumbnailPath(), *it);
+	// If list is already populated, return.
+	if (mGrid.getEntryCount() != files.size()) {
```
What is the problem ?

zigurana replied:
I am mostly just surprised that we *have* to check whether or not to populate the list at all. The populateList() is only called internally, right? Upon construction and with a cursor change (might come from another system, or in a subfolder). So the caller should have a pretty good idea if the call is warranted or not. Is it such a costly methods that we need to guard against too many calls?

Koerty replied:
I'm not going to lie I don't know the reason behind this. I tried to remove the if and do some testing, I tried to search for where this method is called and what method it call. The only thing I'm sure about this is I don't know the reason

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -20,27 +88,64 @@ FileData* GridGameListView::getCursor()
+
+	mGrid.update(deltatime);
+
+	if (mSystemQuickChangeCoolDown < 30)
```
I removed a part of this feature. In the function to handle inputs I removed that code to move it up in the mother class ISimpleGameListView: 
```

	// Quick system change
	if (config->isMappedTo("LeftShoulder", input) || config->isMappedTo("LeftTrigger", input) ||
		config->isMappedTo("LeftTop", input) || config->isMappedTo("LeftBottom", input)) {
		if (Settings::getInstance()->getBool("QuickSystemSelect") && mSystemQuickChangeCoolDown > 10) {
			onFocusLost();
			ViewController::get()->goToPrevGameList();
			return true;
		}
	}
```

it allow the user to always switch systems with LeftShoulder and RightShould whatever game view he is in (basic grid or detailed), that way we don't break his habits if he change the type of game view he use.

I will remove the others references to that var "mSystemQuickChangeCoolDown" since it is unused now.

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -20,27 +88,64 @@ FileData* GridGameListView::getCursor()
+	if (config->isMappedTo("left", input) || config->isMappedTo("right", input)
+	    || config->isMappedTo("up", input) || config->isMappedTo("down", input)) {
+		// Quick and dirty way of changing header title without doing in update()
+		mTitle.setText(mGrid.getSelectedName());
```
I renamed mTitle to mGameTilte and moved the update in the updateInfoPanel function

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -1,16 +1,84 @@
-#include "views/ViewController.h"
+#include "animations/LambdaAnimation.h"
+
+// ===========================================================================
```
Do you mind if I leave that here ? This is pretty well explained I don't want to fck it up with my garbage english

Koerty commented on es-app/src/views/gamelist/GridGameListView.h:
```diff
@@ -1,29 +1,78 @@
 #pragma once
 
 #include "views/gamelist/ISimpleGameListView.h"
+#include "views/gamelist/DetailedGameListView.h"
```
Removed this useless include

Koerty commented on es-core/src/components/GridTileComponent.cpp:
```diff
@@ -0,0 +1,338 @@
+void GridTileComponent::update(int deltaTime) {
+	bool bAnimateChange = false;
+	if (bSelected) {
+		if (mAnimation.frame < mAnimation.maxFrame) {
```
Do you mind doing that change ? I don't know at all how animations work

zigurana replied:
Me neither :-) Let's leave it here for now, and move them in a separate refactoring activity.

Koerty replied:
Maybe I can add a "TODO" comment or open an issue so we don't forget about this.

zigurana replied:
Yeah, adding a TODO tag will suffice, it's not a big deal, certainly not a bug.

Koerty commented on es-core/src/components/GridTileComponent.h:
```diff
@@ -0,0 +1,97 @@
+	unsigned int textColor = 0xFFFFFFFF;
+};
+
+struct CursorAnimation {
```
This is called CursorAnimation because the only animated tile is the one selected by the cursor (and the one which just got unselected when you move the cursor). Should I still change that ?

zigurana replied:
Thanks, I now understand this is only the subset of animations as a result of cursor changes, not the scrolling of the grid itself etc. Your call if you want to keep the name.

Koerty commented on es-core/src/components/IList.h:
```diff
@@ -111,7 +119,7 @@ class IList : public GuiComponent
 	}
 
-	void clear()
+	virtual void clear(bool clearall = false) 
```
Yes you right. I reverted that change

Koerty replied:
clearAll is used by the class ImageGridComponent which inherit from IList and need a bool argument to know if it need to remove all the loaded textures

zigurana replied:
Understood, thanks.

Koerty commented:
@zigurana I did the requested changes and answered most of your questions.

zigurana commented:
@Koerty 
Hey! You were really productive yesterday, good job!
Thanks for taking the time in answering my sometimes stupid questions, and addressing my concerns.
I think that in the revert-activities, you've lost the re-base, so now the PR has some files that should not be part of this review:
* CollectionSystemManager.cpp / .h
* emulationstation.h
* FileData.cpp /.h
* FileFilterIndex.cpp / .h
* SystemData
* SystemScreenSaver
* ...

I think these fake-changes are the result of commit: "[Revert merge](https://github.com/RetroPie/EmulationStation/pull/203/commits/0e55b5fce85b85ea556b6c515d185502060921da)".
If you could repeat the rebase exercise (which should retain all the new commits separately), that would help in seeing what changes you are proposing to make.

I will then have another look at your later commits, as I believe you had some further improvements in mind?
After we've finished the review process, we would like you to squash all the commits into one, so it can be merged in cleanly.

Regarding the MetaData items: I am much in favor of your proposed strategy, so hiding/skipping them when not requested by the theme. Moving them to {1,1} is a hacky workaround for themers, that should not be necessary. 
However, in the current state I believe the components are handled exactly the same as the DetailedGameListView ones, so they will be populated, updated, and shown always. You can leave this in for now, as it is expected behaviour, which will hopefully change with the theming refactoring (I am still working on, sigh).


Again, this is good work, and a significant step up for ES. Thank you.

Koerty commented:
@zigurana

> If you could repeat the rebase exercise (which should retain all the new commits separately), that would help in seeing what changes you are proposing to make.


> I will then have another look at your later commits, as I believe you had some further improvements in mind?

Yes I already have some improvements in mind: adding scroll sound like in textlist (maybe also refactor the way the theme set sounds, check issue #155 for my proposition), completely remake theming syntax, make the grid look better, maybe some performance improvements on how the grid load and unload textures if I find some ...

No problem about squashing I already did it by the past in this PR, please tell me when this PR is ready to be merged and I will do it.

Koerty commented:
@zigurana

> If you could repeat the rebase exercise (which should retain all the new commits separately), that would help in seeing what changes you are proposing to make.

Yes sorry for that I reverted the revert merge.

> I will then have another look at your later commits, as I believe you had some further improvements in mind?

Yes I already have some improvements in mind: adding scroll sound like in textlist (maybe also refactor the way the theme set sounds, check issue #155 for my proposition), completely remake theming syntax, make the grid look better, maybe some performance improvements on how the grid load and unload textures if I find some ... but I will ship them in separate pull request.

> After we've finished the review process, we would like you to squash all the commits into one, so it can be merged in cleanly.

No problem about squashing I already did it by the past in this PR, please tell me when this PR is ready to be merged and I will do it.

> You can leave this in for now, as it is expected behaviour, which will hopefully change with the theming refactoring (I am still working on, sigh).

Ok I leave that to you, good luck :)

I will see tonight or tomorrow if I can answer you lastest questions/changes request, but my work here is nearly done.

Koerty commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -1,16 +1,84 @@
 
 GridGameListView::GridGameListView(Window* window, FileData* root) : ISimpleGameListView(window, root),
-	mGrid(window)
+	mGrid(window, 6/*system->getGridModSize()*/), mBackgroundImage(window), mTitle(window),
```
Removed

Koerty commented:
@zigurana I don't have the answers to your latest question (about the populateList's if and about the "magic numbers). Can we still move forward and finish this PR or do you think this need more work ?

Btw if you want to test the latest version of my GridView branch (where I partially changed grid theming syntax), you can use my fork of @lilbud 's material theme.

Koerty commented:
@zigurana Don't mind my last message. There is so much customization options for the GridTileComponent that I can't find a proper theming syntax. So I reverted the small changes I made on the grid theming syntax and will leave it like this for now.

joolswills commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -49,11 +151,245 @@ void GridGameListView::launch(FileData* game)
+		}
+
+		fadingOut = false;
+	}
```
some indentation issues above

Koerty replied:
Wow thanks for noticing this, my IDE completely fucked up the indentation here for whatever reason it used 4 spaces instead of tabs.

Koerty replied:
This issue is fixed but this comment isn't flagged as "outdated" because you put it on the lines below the problem and not directly on it.

joolswills commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -1,16 +1,86 @@
+	mDescContainer.setSize(mSize.x() * (0.50f - 2 * padding), mSize.y() - mDescContainer.getPosition().y());
+	mDescContainer.setAutoScroll(true);
+    mDescContainer.setDefaultZIndex(40);
+    addChild(&mDescContainer);
```
wrong indentation.

Koerty replied:
Wow thanks for noticing this, my IDE completely fucked up the indentation here for whatever reason it used 4 spaces instead of tabs.

joolswills commented on es-app/src/views/gamelist/ISimpleGameListView.cpp:
```diff
@@ -118,15 +118,15 @@ bool ISimpleGameListView::input(InputConfig* config, Input input)
 
 			return true;
-		}else if(config->isMappedTo("right", input))
+		}else if(config->isMappedTo("right", input) || config->isMappedTo("rightshoulder", input))
```
is this related to this commit ?

Koerty replied:
Changing system while in game view is usually done with right and left arrows. In the grid view we can't do that since left and right arrows control the selector (along with up and down) so switching system is done with left and right shoulder. For not breaking user habits I decided to allow switch system with left and right shoulder for all the game view. If later we decide to allow the user to choose different game view style depending on the system it will also be really handly (for example an automatic option which display the grid view for systems with metadata and the basic view for systems without metadata, like the automatic option we have right know which choose between detailed and basic view with the same criterias)

Koerty commented:
@joolswills Is your review over ?

zigurana commented on data/converted/folder_png.cpp:
```diff
@@ -0,0 +1,102 @@
+// this file was auto-generated from "folder.png" by res2h
```
Idea: maybe make this folder image greyscale, so that themers can colorize them to their wishes via the `<color>` tag?

Koerty replied:
Your wish is my command

Koerty replied:
I wasn't sure how to properly allow themes to change icon color so I added a tag to completely change the icon, see my last comment for more informations.

zigurana commented on es-app/src/views/gamelist/GridGameListView.cpp:
```diff
@@ -20,27 +90,59 @@ FileData* GridGameListView::getCursor()
+	if(!mGrid.setCursor(file) && mGrid.getEntryCount() > 0)
 	{
-		populateList(file->getParent()->getChildrenListToDisplay());
+		populateList(file->getParent()->getChildren());
```
I think you will need to call `getChildrenListToDisplay()`, to support filtering of entries. Why would you need to call `getChildren()` directly?

Koerty replied:
Thanks for noticing this ;)

zigurana commented on es-core/CMakeLists.txt:
```diff
@@ -171,6 +175,9 @@ set(EMBEDDED_ASSET_SOURCES
     ${emulationstation-all_SOURCE_DIR}/data/converted/busy_3_svg.cpp
+
+    ${emulationstation-all_SOURCE_DIR}/data/converted/blank_game_png.cpp
+	${emulationstation-all_SOURCE_DIR}/data/converted/folder_png.cpp
```
whitespace?

zigurana commented on es-core/src/components/GridTileComponent.cpp:
```diff
@@ -0,0 +1,338 @@
+	if (elem) {
+		if (elem->has("color")) {
+			mAnimation.selected.color = elem->get<unsigned int>("color");
+			//mAnimation.selected.color = addAlpha(mAnimation.selected.color, 0xFF);
```
this comment can go?

zigurana commented on es-core/src/components/GridTileComponent.cpp:
```diff
@@ -0,0 +1,338 @@
+	if (elem) {
+		if (elem->has("color")) {
+			mAnimation.unselected.color = elem->get<unsigned int>("color");
+			//mAnimation.unselected.color = addAlpha(mAnimation.current.color, 0xAA);
```
idem dito

jacobfk20 commented:
@Koerty Keep up the awesome work!  I haven't been on github in so long and I was glad to see that someone was able to pick up where I left off.  I was also happy to see that the retropie fork of ES has a ton activity as well.  I'll try to keep an eye on this branch from here on out. I'd also like to thank you for continuing were I left off and I hope my code wasn't too much of a mess for ya ;)

dudleydes commented:
I have been using jacobfk20's Gridview mod since the end of the last year so I thought I'd check out this PR.

First of all, excellent work! 

I did have to rework my theme a little. The main issue was the z-index feature which was introduced since the Gridview mod I was using was built and is unrelated to this PR. There were some other minor issues which I had to tweak.

Thanks for your hard work. It's really appreciated to finally able to have this feature in the main ES branch.

Koerty commented:
The legendary @jacobfk20 answered my pull request, I will never wash my computer screen ever again.

Joke appart, thank you for your message I trully appreciate that. Regarding the code I find it pretty clear and well structured it's a pleasure to work on it. My only concerns are about some "magic numbers", like the range around the cursor's index used for loading textures which is set to 12 and I can't explain why, I can only guess it was first made for a 4 columns and 3 rows grid and never changed after that.

jacobfk20 commented:
@koerty I can't really remember why but it would likely be from testing to see the exact number of tiles could be loaded into vram at a given time and seeing how far it could be pushed.  Over time it would have been changed to be set by another function that would either determine the size based on grid rows and columns or by texture sizes. 

If I have time this weekend I'm going to build this branch and check it out.  I also noticed there was a change in ES theme options that could allow better control over grid view.  I remember everything just being assigned to <ninepatch> because <image> and <container> only had size parameters and adding anything else would break compatibility with vanilla ES.  It made themeing options for grid view pretty hard to do.

lilbud commented:
Any update on this?

PokeEngineer commented:
Well @lilbud, there is a merge conflict that needs to be solved, and @Koerty hasn't been active on GitHub for about a month...

lilbud commented:
@PokeEngineer And gridview dies again, good night everybody!

PokeEngineer commented:
@lilbud 

I wonder if the concept of GridView in ES is cursed, because everyone who picks it up is really active for awhile, and then they just stop visiting GitHub... (Obviously I'm joking, but it is still strange...)

lilbud commented:
@PokeEngineer Its not just Github, after a while they go silent. On the forums, here, etc.

PokeEngineer commented:
@lilbud 

Oh yeah... Forgot about that...

jacobfk20 commented:
Gridview will be finished eventually.  It is a huge project to take on!  When I was working on it I remember always thinking that EmulationStation itself needed a ton of work to make things easier for Gridview to really take off.  ES used to have a problem with creating your own custom theme parameters so I wrote the theme options around this limitation, which really made things complicated for theme makers...
Another thing that can slow development is how popular this feature is.  Many people come together to help lay the ground work for how themeing should be done and how everything should be handled in general and this can pretty overwhelming sometimes for one person to handle :)
Also, life gets in the way or changes sometimes ;)

Don't lose hope.

PokeEngineer commented:
@jacobfk20 

Yeah, I know... ☺

Koerty commented:
@joolswills Hey I'm trying to merge lastest emulationstation changes into my branch and there is something I don't understand. Was there any change about boost filesystem that I missed ? I got some errors now that I didn't have by the past, about the functions "boost::filesystem::exists" and "boost::filesystem::remove"

`/home/koerty/EmulationStation/es-app/src/views/gamelist/GridGameListView.cpp: In member function ‘virtual void GridGameListView::remove(FileData*, bool)’:
/home/koerty/EmulationStation/es-app/src/views/gamelist/GridGameListView.cpp:158:22: error: ‘remove’ is not a member of ‘boost::filesystem’
   boost::filesystem::remove(game->getPath());  // actually delete the file on the filesystem
`

Koerty commented:
@zigurana I've properly merged emulationstation lastest changes so we can finish this pull request. I took a 2nd look about your lastest issue (themer should be able to change folder color) and I'm still not sure how to properly implement this. Beside that and the issue right above with boost::filesystem, what is the next step ? Do you still have more verifications, testing, reviewing ... planned ?

zigurana commented:
Hey Koerty! Good to see you're back working on this. A lot has happened to the codebase in the last few months, especially with changes by @tomaz82. I'll try to have a look at the current state of this PR asap.

hex007 commented:
I just checked on this PR yesterday and thought, did another dev who worked on Gridview went missing. Nice to see that the curse is not that true.

Koerty commented:
@zigurana I wasn't sure how to properly change folders icon's color (either with animations or ImageComponent) so I added a tag for themes to changes the folder icon using the following syntax : 
```xml
<view name="grid">
	<image name="folder">
		<path>./art/folder.png</path>
	</image>
</view>
```
If this is ok for you I think this should do the trick (at least for now)

Koerty commented on es-core/src/components/ImageGridComponent.h:
```diff
@@ -1,15 +1,35 @@
+// A range around the cursor's index used for loading in textures
+struct CursorRange {
+	int min = 0;
+	int max = 12;
```
I think it's set by default to 12 because we used a 3 * 4 grid at this time. It is dynamically adjusted at the end of the function updateLoadRange() anyway.

Koerty commented:
Everyone seems a bit lost about the current statue of the gridview (including me). So I made this "small" roadmap of future gridview developments. Feel free to tell me if I missed something.

SHORT TERM (before this PR is merged)
- ??

--------- 
LONG TERM (future improvements for the grid view shipped in different pull requests)
- Allow themes to change folder icon color (as suggested by @zigurana) For now they can completly change the icon (see comment right above).
- Move animation code from GridTileComponent > update() to AnimationUtil.cpp (as suggested by @zigurana)
- Add scrolling sound (just like in TextListComponent)
- Refactor theming syntax (+ merge GRIDTHEME.md into THEMES.md) : see the discussion about theming syntax https://github.com/RetroPie/EmulationStation/issues/206
- Change grid view display code : see end of my comment on the discussion about theming syntax here : https://github.com/RetroPie/EmulationStation/issues/206#issuecomment-326805924 + other comments from @jrassa about stretching images which have different aspect ratio
- Refactoring/Optimizations if I find some

Koerty commented:
@zigurana Do you think you will have time to check this anytime soon ? So we can move forward and finish this.

Koerty commented on es-core/src/components/GridTileComponent.h:
```diff
@@ -0,0 +1,97 @@
+	bool isSelected() { return bSelected; }
+
+	// Used to give custom theme control to all elements and for slight animation puroposes.
+	void setTheme(const std::shared_ptr<ThemeData>& theme);
```
This have been changed in a past commit but I forgot to write it as done here, so ... Done.

PokeEngineer commented:
@Koerty 

Tagging @joolswills because he might be helpful here too.

Koerty commented:
@joolswills You started a review a while ago but it's not showing up as "approved". Are you done reviewing this pull request ?

Koerty commented:
@joolswills @zigurana @jrassa Is there anybody here ?

joolswills commented:
I've not reviewed it, but I am not familiar enough with this part of the codebase really. I would be interested in what @tomas82 thinks.

joolswills commented:
Correction. I made some review comments only earlier on. I think they were cosmetic ones that may be already sorted.

Koerty commented:
@joolswills Ok just wanted to be sure since your review isnt showing up as "approved" like jrassa's review.

I have no news from @zigurana since mid december so if @tomaz82 want to start taking a look at this PR he is welcome.

jrassa commented:
@Koerty the "approved" label shows up for me, because I marked a specific review I did as approved, but it wasn't an exhaustive review, and I can not say for sure if I approve this PR in its current state.  I see now there is a way to remove the "approved" label and I will be doing that.

As I have stated previously, I feel that the theme syntax (#206) should be figured out before this PR is accepted.  It is not my decision though and if others deem it good enough it may be accepted.

Koerty commented:
@jrassa I thought you said you didn't mind seeing it come in a future pull request but I can't find your message. My bad then, let's do this refactoring and move forward. I updated the todo list (9 comments ago) and will take a look at this in the next days.

tomaz82 commented:
I agree with @jrassa, not necessarily for the same reasons. Looking at this from a more professional perspective, it's "dangerous" to "release" something that's not finalized into the wild, such as the current theming sytax for gridview, because once it's out, theme authors will start implementing gridview, using a syntax that might be, I shouldn't sa y bad, but, not as good as it could be, making it hard to fix the syntaxing afterwards.

I would personally also split this up into smaller PR's where possible, like as example, I'd make a patch for allowing using left/right shoulder buttons to switch systems, since that "works" in non-gridview as well. Basically break this down into pieces, isolating anything that isn't gridview specific, so that the final gridview PR will be as condensed as possible.

Pardon my wordings, and if something doesn't make sense, english isn't my native language.

Koerty commented:
@tomaz82 Two last things about theming syntax : 
- there already are themes out here that support grid view, like the Material theme from lilbud or this SNES theme jrassa talked about in the issue I opened, if we change theming syntax before closing this pull request we will break this themes (there isn't much of them but still)
- We will still do small changes to theming syntax in the futur, For example @jrassa asked for defining the grid's tiles size using only the grid size and tile max size, instead of grid size, grid number of row and columns and grid margins as it is now

Your move if you want to change theming syntax now, I'm not against it, I'm just trying to find counter arguments in case you didn't thought about it.

About splitting this PR in smaller ones : 
- I think we may commit the animation stuff in another PR even if it's only used here -> +97 lines over 3 files 
- also the newly added ressources (folder image and blank game image) -> +1545 lines -27 lines over 7 files (blank_game_png.cpp take 1438 lines alone)

> Edit: And I'm not sure if this has changed, but earlier comments here suggest there is still code in this PR that is not entirely known what it does, or why it does what it does, which is always a risk, merging code that noone knows how it works.

In my recent changes (on 17 Dec 2017) I took a more indepth look into the code, trying to understand what I didn't understand before. I won't lie to you there is still one or two things I don't understand about the dynamic image loader. The problem for me is if I modify it right now and break something or introduce a new bug, it will be hard to figure out why or reverse the changes since it will be in the same PR.

tomaz82 commented:
@Koerty
I see your point about the theming syntax, but there is a big difference to "break" those 2 themes, compared to having gridview in the official retropie build, and have dozens of themes adding support for it, and then need to change the syntax. From experience, it's a bad idea to release a new feature until it's "done".

I didn't read through the entire thing, I just saw bits and pieces that could be split out of this pr, and now that you meantion it, the animation thing, not sure why it's called animation, it seems like it's doing color operations, sure it's used by gridview transitions, but I don't see why it's in a class called Animation.

It's great that you now know more of the code, that's one of the 2 goals when working on something like this, #1 bring something for others to enjoy, #2 learn new stuff. But the fact that there still are unknown things in the code worries me.

If this is not completed by the time I'm done with all my ongoing PRs I can take a stab at this as well and see if we can figure out the last bits and pieces together.

Koerty commented:
@jrassa @joolswills If your concern is to only see the grid view on EmulationStation's main branch only when it's completly finished and 200% stable then maybe we should create a "GridView" branch on this repository so I can send my pull requests to this branch, and later on when you will be satisfied from it we will merge it in the main branch.

This would allow you to monitor carefully all my pull requests, and allow me to ship each big changes (refactors on theming syntax, grid tiles display, dynamic image loading, ....) in different pull requests for easier reviewing and reverting/understanding.

Koerty commented:
@tomaz82 

> the animation thing, not sure why it's called animation, it seems like it's doing color operations, sure it's used by gridview transitions, but I don't see why it's in a class called Animation

The AnimationUtil class is an utility class used by the grid's tiles to do text color shifting when the tile is selected or unselected (basically changing from unselected color to selected color over X frames).

What about ressources (adding blank game image and folder image) in a separate pull request ? Should I do that ?

Koerty commented:
> As I have stated previously, I feel that the theme syntax (#206) should be figured out before this PR is accepted. It is not my decision though and if others deem it good enough it may be accepted.

I made a new theming syntax proposition in the related issue #206

Koerty commented:
@tomaz82 I updated this PR accordingly to the huge changes introduced by #361 and #367 

Do you have time to check this PR anytime soon or you still have many upcoming PR ?

Koerty commented:
Have everyone gave up on this pull request or what is happening ?

PokeEngineer commented:
@Koerty 

I, sure the heck didn't!

lilbud commented:
I've said it before, its the curse of gridview.

PokeEngineer commented:
@lilbud 

<sub><sup>I said that as a joke...</sup></sub>

Koerty commented:
@PokeEngineer @lilbud Stop please, I already told you this is mean.

joolswills commented:
This issue tracker is not the place for silly replies. If you don't have a comment related to the code, don't post. Use the forum for chat.

joolswills commented:
It will be reviewed when people are free. Everyone is probably busy currently. Thanks.

Koerty commented:
@joolswills Thanks for your answer. If that can help I can completly disassemble the GridView in really small piece and ship them separatly. Make a first PR with only a really basic GridView, which just display the images of the games, then ship the rest in different pull requests (dynamic image loader, animations, ...). A bit like @tomaz82 suggested some messages ago but even more splitted.

That would give me way more works but if it's easier and faster to review after that, since you don't have much free time, I'm ready to do it.

hex007 commented:
@Koerty  Has the theming syntax and other design considerations finalized ?

I agree splitting the implementation into smaller batches of commits (releases) would be easier to test and debug. Now that the code is somewhat stable, breaking the implementation into parts would be easier too. I am up for testing. Let me know what all is needed in form of feedback.

PokeEngineer commented:
@Koerty, @joolswills 

Sorry, I didn't mean to be mean... I'll just remove my self out of here...

Koerty commented:
@hex007 I'm not sure for the theming syntax, I got zero feedback from main RetroPie's contributors after my last theming syntax proposition : https://github.com/RetroPie/EmulationStation/issues/206#issuecomment-361534635

Koerty commented:
This is the new plan for the grid view, each point will be shipped in a separate pull request for easier reviews.

In the "preparation" step, the grid view will not be ready for users to use it, so I can either enable it at the PR creation and disable it right before the PR is merged, or just hide it from the "Gamelist view style" option in the menu until it's ready. This is so reviewers can easily try it out before it's completely ready.

PREPARATION
- Allow swapping between systems using left shoulder and right shoulder
- Rewrite the grid view display code written by Aloshi (using a "tileSize" parameter from theme, suggested by @jrassa on https://github.com/RetroPie/EmulationStation/issues/206#issuecomment-326809971)
- Add metadata labels and values (sames as in the detailed view)
- Add a new metadata label and value : game title
- Add new default icon for games without cover
- Add new default icon for folders

RELEASE
- Enable the grid view for users

AMELIORATION
- Add game title under game cover
- Add dynamic image loader
- Add more animations
- Add navigation sound

The new theming syntax will be used https://github.com/RetroPie/EmulationStation/issues/206#issuecomment-361534635

I hope I didn't go too far in breaking this PR in smaller parts. Please let me know if you have any question or suggestion, and if it's ok for you then tell me and I will start right now releasing the first PRs.

hex007 commented:
I would suggest hiding the gridview in settings. So only those with manual edits to config file can see gridview

pjft commented:
@hex007 what do you mean? During the rollout, or afterwards?

@koerty I'll take a stab at looking through this in the coming days/weeks when I have some more time. Thanks.

tomaz82 commented:
I'm at GDC so I'm insanely busy right now, not sure when I'll get any free time that I'll be able to dedicate to RetroPie, but I'll try.

hex007 commented:
@pjft  During the development, for testing purposes. Once its stable doors can be opened for themers and users

Koerty commented:
@pjft @tomaz82 I'm trying to push this feature more "aggressively" so it don't sit here for another year, but it can wait a few days or weeks if you are too busy right now, no problem with that.

Koerty commented:
Small note for those who aren't actively following this github repository.

The grid view have been completely disassembled and rewrote from scratch, each part is getting carefully reviewed and merged into the ES master branch.

This mean a lot of code have changed, and as it would be cool to keep this PR updated to see what features are missing, it is way too much effort due to the considerable amount of changes made on the road.

I keep this PR open until we reach the V1 of the grid view really soon, then I will take a deeper look into its code and comments to be sure I didn't miss anything, and then close it.

Koerty commented:
I'm closing this, as the base grid view features have reached the master branch of this repository, and the others have been added to the todo list of the grid view wiki.

There is still a few features in this PR that have not been implemented yet (game name in the grid, selected tile size and color changes animation, ...) so I will not delete this branch anytime soon.
